### PR TITLE
Re-introduce code coverage

### PR DESCRIPTION
**Pull request by Sybrand**
# Overview
Re-introduce code coverage and speed up test runs (running about x2 faster now).

# Details
- run web and api integration in parallel - save results as artifacts.
- upload coverage reports to codecov - use artifacts.
- add tests links to pr description if none present.
- fixed up R to use cache.
- moved linting to run sooner (so we can fail faster).

# Test Links:

# Test Links:
[Percentile Calculator](https://wps-pr-1379.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-1379.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-1379.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-1379.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-1379.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=1108&f=c5&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN#state=2ec784ca-c46a-49d0-b2b3-1cf32a9015a2&session_state=7d9447c8-db66-4661-b4cb-03d2ac0d1d8f&code=32292df4-2bdf-4f90-a4a8-c8dbcda682a9.7d9447c8-db66-4661-b4cb-03d2ac0d1d8f.2b63f390-f3dc-43ae-89f2-016453863476)
[HFI Calculator](https://wps-pr-1379.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: .github/workflows/integration.yml [hunk 1]
```diff
@@ -89,6 +55,42 @@ jobs:
...
           restore-keys: |
             ${{ runner.os }}-poetry-cache-
+      - name: Install python dependencies using poetry (api)
+        working-directory: ./api
+        run: |
+          source ~/.poetry/env
+          poetry install
+      - name: Lint (api)
+        # We run lint before unit tests, and before installing some of the
+        # longer running dependancy installs because it runs faster, so we'll reach a fail condition sooner.
```
Comment (conbrad) [inline .github/workflows/integration.yml:65]: dependancy -> dependency

File: .github/workflows/integration.yml [hunk 2]
```diff
@@ -126,40 +117,73 @@ jobs:
...
+          path: ./web/finalCoverage
+  upload-code-coverage:
+    name: Coverage with CodeCov
+    runs-on: ubuntu-20.04
+    needs: [lint-and-test-web, lint-and-test-api]
+    steps:
+      # we need to checkout, so that we have codecov.yml
+      - name: Checkout repo
+        uses: actions/checkout@v2
+      - name: Download all workflow run artifacts
```
Comment (conbrad) [inline .github/workflows/integration.yml:182]: Was it possible to verify that all coverage reports were downloaded and posted to CodeCov?

File: .github/workflows/pr_description.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  const prBaseUrl = `https://wps-pr-${prNum}.apps.silver.devops.gov.bc.ca`;
+  if (pr.data.body === null || !pr.data.body.includes(prBaseUrl)) {
+    // If the body doesn't already contain some test links, we create a few.
+    let body = pr.data.body === null ? "" : pr.data.body;
+    body += "\n# Test Links:\n";
+    body += `Percentile Calculator: ${prBaseUrl}/\n`;
+    body += `MoreCast: ${prBaseUrl}/morecast\n`;
+    body += `C-Haines: ${prBaseUrl}/c-haines\n`;
+    body += `FireBat: ${prBaseUrl}/fire-behaviour-calculator\n`;
+    body += `FireBat bookmark: ${prBaseUrl}/fire-behaviour-calculator?s=264&f=o1a&c=6&w=undefined,s=322&f=c1&c=undefined&w=15.5\n`;
```
Comment (conbrad) [inline .github/workflows/pr_description.js:20]: We should use the larger, realistic lists like the Kamloops one (from my browser history, may be different, but its large): 

```
fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=1108&f=c5&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN#state=2ec784ca-c46a-49d0-b2b3-1cf32a9015a2&session_state=7d9447c8-db66-4661-b4cb-03d2ac0d1d8f&code=32292df4-2bdf-4f90-a4a8-c8dbcda682a9.7d9447c8-db66-4661-b4cb-03d2ac0d1d8f.2b63f390-f3dc-43ae-89f2-016453863476 
```

**Review by conbrad — APPROVED** [review]
Nice 😎

**Review by JulianForeman — APPROVED** [review]
